core/os/android/adb/commands.go

```diff
@@ -256,6 +295,47 @@ func (b *binding) SupportsPerfetto(ctx context.Context) bool {
 	return os.GetAPIVersion() >= 28
 }
 
+func (b *binding) QueryPerfettoServiceState(ctx context.Context) (*device.PerfettoCapability, error) {
+	if b.Instance().GetConfiguration().GetOS().GetAPIVersion() <= 29 {
+		return &device.PerfettoCapability{}, log.Errf(ctx, nil, "Querying perfetto capability requires Android API > 29\n")
+	}
+	res, err := b.Shell("perfetto", "--query-raw", "|", "base64").Call(ctx)
+	if err != nil {
+		return &device.PerfettoCapability{}, log.Errf(ctx, nil, "adb shell perfetto returned error: \n%s", res)
+	}
+	decoded, _ := base64.StdEncoding.DecodeString(res)
+	tracingServiceState := &perfetto_pb.TracingServiceState{}
+	err = proto.Unmarshal(decoded, tracingServiceState)
+	if err != nil {
+		return &device.PerfettoCapability{}, log.Errf(ctx, nil, "Unmarshal returned error: \n")
+	}
+	perfettoCapability := &device.PerfettoCapability{
+		GpuProfiling: &device.GPUProfiling{
+			HasRenderStage:       false,
+			GpuCounterDescriptor: nil,
+		},
+	}
+	dataSources := tracingServiceState.GetDataSources()
+	for _, dataSource := range dataSources {
+		dataSourceDescriptor := dataSource.GetDsDescriptor()
+		if dataSourceDescriptor.GetName() == "gpu.renderingstages" {
+			perfettoCapability.GpuProfiling.HasRenderStage = true
+			continue
+		}
+		gpuCounterDescriptor := dataSourceDescriptor.GetGpuCounterDescriptor()
+		if gpuCounterDescriptor == nil {
```

Comment:
Checking for nil with protos goes against established patterns and is not necessary. All the `Get*` functions hide the "nilness". You can simply remove this check and `gpuCounterDescriptor.GetSpecs()` will return an empty list if `gpuCounterDescriptor` is nil.